File: MorseApp/app/src/main/java/first/app/morseapp/Dictionary.java
package first.app.morseapp;
/**
 * A class, that reads in the morse.xml file and stores the values in a hashmap.
 * One way from a normal character, i.e. 'a', to it's respective morse encoding
 * and the other way around.
 *
 * 10.08.2022
 */

import android.util.Xml;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public class Dictionary {

    private HashMap<String, String> dictAToM;
    private HashMap<String, String> dictMToA;


    public Dictionary(){
        dictAToM = new HashMap<>();
        dictMToA = new HashMap<>();
    }
    /**
     * Initializes the dictionary for the morse code
     * @param in the morse.xml to read in with an input stream
     * @throws IOException if file not found
     */
    public void initializeDic(InputStream in) throws IOException, XmlPullParserException{
        parseMaps(in);
    }

    /**
     * Get the value of an alphabet key. i.e. a : .-
     * @param key the character to look for
     * @return the respective value, if not inside dictionary, return not found
     */
    public String getId(String key){
        if(dictAToM.containsKey(key.toLowerCase()))
            return dictAToM.get(key.toLowerCase());

        return "[Not found]";
    }

    /**
     * Get the value of a morse encoding. i.e. .- : a
     * @param key the morse endocing to look for
     * @return the rsspective decription, or not found if not in dicitonary
     */
    public String getMorse(String key){
        if(dictMToA.containsKey(key))
            return dictMToA.get(key);

        return "[Not found]";
    }
    /**
     * helper method to parse the dictionary
     */
    private void parseMaps(InputStream in) throws IOException, XmlPullParserException {
        try {
            XmlPullParser reader = Xml.newPullParser();
            reader.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
            reader.setInput(in, null);
            reader.nextTag();
            parseMorse(reader);
        } finally {
            in.close();
        }
    }
    // a message from my girlfriend
    //hallooooooooooooooo


    /**
     * helper method to parse the id and morse of a token
     * @param parser the xml event parser parsing the document
     */
    private void parseMorse(XmlPullParser parser) throws XmlPullParserException, IOException {
        while(parser.next() != XmlPullParser.END_DOCUMENT) {
            if (parser.getEventType() == XmlPullParser.START_TAG) {
                String name = parser.getName();
                if (name.equals("letters"))
                    readLetters(parser);

                else if (name.equals("numbers"))
                    readNumbers(parser);
            }
        }
    }

    /**
     * helper method to parse the letters in the xml file
     * @param parser the xml pull parser
     * @throws XmlPullParserException if something goes wrong while parsing
     * @throws IOException if something goes while parsing
     */
    private void readLetters(XmlPullParser parser) throws XmlPullParserException, IOException {
        parser.require(XmlPullParser.START_TAG, "", "letters");
        // boolean to check whether the end of the letters tag is reached
        boolean looper = true;
        while(looper){
            if(parser.getEventType() == XmlPullParser.START_TAG && parser.getName().equals("letter"))
                readLetter(parser);

            if(parser.getEventType() == XmlPullParser.END_TAG && parser.getName().equals("letters"))
                looper = false;

            parser.next();
        }
    }

    /**
     * helper method to read in a single letter in the letters tag
     * @param parser the xml pull parser
     * @throws XmlPullParserException if something goes wrong while parsing
     * @throws IOException if something goes wrong while parsing
     */
    private void readLetter(XmlPullParser parser) throws XmlPullParserException, IOException {
        if(parser.getEventType() == XmlPullParser.START_TAG && parser.getName().equals("letter"))
            readToken(parser);
    }

    /**
     * helper method to read in the numbers part of the xml file
     * @param parser the xml pull parser
     * @throws XmlPullParserException if something goes wrong while parsing
     * @throws IOException if something goes wrong while parsing
     */
    private void readNumbers(XmlPullParser parser) throws XmlPullParserException, IOException {
        parser.require(XmlPullParser.START_TAG, "", "numbers");
        // boolean to check, whether the end of the numbers tag is reached
        boolean looper = true;
        while(looper){
            if(parser.getEventType() == XmlPullParser.START_TAG && parser.getName().equals("number"))
                readNumber(parser);
            if(parser.getEventType() == XmlPullParser.END_TAG && parser.getName().equals("numbers"))
                looper = false;

            parser.next();
        }
    }

    /**
     * helper method to read in the number part of numbers
     * @param parser the xml pull parser
     * @throws XmlPullParserException if something goes wrong while parsing
     * @throws IOException if something goes wrong while parsing
     */
    private void readNumber(XmlPullParser parser) throws XmlPullParserException, IOException {
        if(parser.getEventType() == XmlPullParser.START_TAG && parser.getName().equals("number"))
            readToken(parser);
    }

    /**
     * the helper method to read in each token of the tags
     * @param parser the xml pull parser
     * @throws XmlPullParserException if something goes wrong while parsing
     * @throws IOException if something goes wrong while parsing
     */
    private void readToken(XmlPullParser parser) throws XmlPullParserException, IOException {
        // boolean to determine end of token.
        // for reasons which must be cleared yet, to get into the actual token part.
        boolean go = true;
        while(go){
            if (parser.getEventType() == XmlPullParser.START_TAG && parser.getName().equals("token")) {
                String id = parser.getAttributeValue(null, "id");
                String morse = parser.getAttributeValue(null, "morse");
                dictAToM.put(id, morse);
                dictMToA.put(morse, id);
            }
            parser.next();
            if(parser.getEventType() == XmlPullParser.END_TAG && parser.getName().equals("token"))
                go = false;
        }
    }
}

File: MorseConverterKt/src/main/java/ap/PhilippHellwig/morseconvertkt/Dictionary.kt
/**
 * this class creates the morse hash maps and allows to retrieve the
 * value of either the character key or the morse key.
 * i.e. .- = a
 * or a = .-
 */
package ap.PhilippHellwig.morseconvertkt

import android.util.Xml
import org.xmlpull.v1.XmlPullParser
import java.io.InputStream

class Dictionary() {

    private lateinit var dictAToM : HashMap<String, String>
    private lateinit var dictMToA : HashMap<String, String>

    /**
     * initializes the hashmaps
     * @param input the morse xml stream
     */
    fun initializeDic(input : InputStream){
        dictAToM = HashMap()
        dictMToA = HashMap()
        parseMaps(input)
    }

    /**
     * get the value of a character key
     * @param key the character key, i.e. a
     */
    fun getID(key: String) =
        dictAToM.getOrElse(key.lowercase()){"[Not found]"}


    /**
     * get the value of a morse key
     * @param key the morse sign key, i.e. .-
     */
    fun getMorse(key: String) =
        dictMToA.getOrElse(key){"[Not found]"}

    /**
     * helper function to initiate the parsing
     * @param input the morse xml stream
     */
    private fun parseMaps(input: InputStream){
        val reader = Xml.newPullParser()
        reader.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
        reader.setInput(input, null)
        reader.nextTag()
        parseMorse(reader)
    }

    /**
     * helper function to parse the morse xml
     * @param reader the parser for the morse xml
     */
    private fun parseMorse(reader: XmlPullParser) {
        while (reader.next() != XmlPullParser.END_DOCUMENT){
            if(reader.eventType == XmlPullParser.START_TAG) {
                var name: String = reader.name
                if(name.equals("letters"))
                    readLetters(reader)

                else if(name.equals("numbers"))
                    readNumbers(reader)
            }
        }
    }

    /**
     * helper function to read the letters of the xml file
     * @param reader the parser for the morse xml
     */
    private fun readLetters(reader: XmlPullParser){
        reader.require(XmlPullParser.START_TAG, "", "letters")
        var looper = true
        while(looper){
            if(reader.eventType == XmlPullParser.START_TAG && reader.name.equals("letter"))
                readLetter(reader)
            if(reader.eventType == XmlPullParser.END_TAG && reader.name.equals("letters"))
                looper = false

            reader.next()
        }
    }

    /**
     * helper function to read the letter member
     * @param reader the parser for the morse xml
     */
    private fun readLetter(reader: XmlPullParser){
        if(reader.eventType == XmlPullParser.START_TAG && reader.name.equals("letter"))
            readToken(reader)
    }

    /**
     * helper function to read the the numbers of the xml file
     * @param reader the parser for the morse xml
     */
    private fun readNumbers(reader: XmlPullParser){
        reader.require(XmlPullParser.START_TAG, "", "numbers")
        var looper = true
        while(looper){
            if(reader.eventType == XmlPullParser.START_TAG && reader.name.equals("number"))
                readNumber(reader)
            if(reader.eventType == XmlPullParser.END_TAG && reader.name.equals("numbers"))
                looper = false
            reader.next()
        }
    }

    /**
     * helper method for the number member
     * @param reader the parser for the morse xml
     */
    private fun readNumber(reader: XmlPullParser){
        if(reader.eventType == XmlPullParser.START_TAG && reader.name.equals("number"))
            readToken(reader)
    }

    /**
     * reads the token of the current member
     * @param reader the pull parser to read the morse xml
     */
    private fun readToken(reader: XmlPullParser){
        var go = true
        while(go){
            if(reader.eventType == XmlPullParser.START_TAG && reader.name.equals("token")){
                var id = reader.getAttributeValue(null, "id")
                var morse = reader.getAttributeValue(null, "morse")
                dictAToM[id] = morse
                dictMToA[morse] = id
            }
            reader.next()
            if(reader.eventType == XmlPullParser.END_TAG && reader.name.equals("token"))
                go = false
        }
    }
}